### ISSUE TITLE
test

### ISSUE THREAD

**Issue by yingkewei** [x-xc/SACB_Net#4]
Dear author , could you provide test.py?

**Comment by x-xc** [comment]
import glob
import os, utils
from torch.utils.data import DataLoader
from data import datasets, trans
import numpy as np
import torch
from torchvision import transforms

import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import axes3d
from natsort import natsorted
#from digital_diffeomorphism.ndv import cal_ndv
from surface_distance import *
import random
from medpy.metric import binary
import utils 
from medpy import metric
os.environ["base_dir"] = '/bask/projects/d/duanj-ai-imaging/xxc/dataset_all'

labels = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]
def assd_cal(y_pred, y_true, x_true, labels, vspace= np.ones(3)):
    assd  = 0
    count = 0
    for i in labels:
        if ((y_true==i).sum()==0) or ((x_true==i).sum()==0):
            continue
        assd += metric.assd(y_pred == i, y_true == i, voxelspacing=vspace)
        count += 1
    return assd/count

def hd95_cal(y_pred, y_true, x_true, labels, vspace= np.ones(3)):
    hd95  = 0
    count = 0
    for i in labels:
        if ((y_true==i).sum()==0) or ((x_true==i).sum()==0):
            continue
        hd95+=metric.hd95(y_pred == i, y_true == i, voxelspacing=vspace)
        count += 1
    return hd95/count

def main():

    weights = [1, 0.3] 
    lr = 1e-4
    model_idx = - 1
    k = 9

    tag = f'tune_abd_model3s4_c4_k{k}_scale4'
    from model import SACB_Net
    
    model_folder = 'sacb_abd_ncc_{}_reg_{}_{}/'.format(weights[0], weights[1], tag)
    model_dir = 'experiments/' + model_folder

    line = 'dice' + ',' +  'j' + ',' + 'hd95' + ',' + 'assd' + ',' +'ndvm' + ',' + 'ndvmp' + ',' + 'ndv'+ ',' + 'ndvp'
    csv_writter(line,'Quantitative_Results/' + model_folder[:-1])
    
    test_dir = os.path.join(os.getenv('base_dir'), 'AbdomenCTCT/Test/')
    img_size = (192, 160, 224)

    '''
    Initialize model
    '''
    model = SACB_Net(inshape=img_size, num_k=k)
  
    pth = natsorted(glob.glob(model_dir+'*.pth.tar'))[model_idx]
    best_model = torch.load(pth)['state_dict']
    print('Best model: {}'.format(pth))
    model.load_state_dict(best_model)
    model.cuda()
    model.set_k(k)
    
    reg_model = utils.register_model(img_size, 'nearest')
    reg_model.cuda()
    test_composed = transforms.Compose([trans.NumpyType((np.float32, np.int16))])
    test_set = datasets.LPBABrainInferDatasetS2S(sorted(glob.glob(test_dir + '*.pkl')), transforms=test_composed)
    test_loader = DataLoader(test_set, batch_size=1, shuffle=False, num_workers=0, pin_memory=True, drop_last=True)
    eval_dsc_def = utils.AverageMeter()
    eval_dsc_raw = utils.AverageMeter()
    eval_det = utils.AverageMeter()
    with torch.no_grad():
        stdy_idx = 0
        for data in test_loader:
            model.eval()
            data = [t.cuda() for t in data]
            x = data[0]
            y = data[1]
            x_seg = data[2]
            y_seg = data[3]
            output, flow = model(x, y)
            def_out = reg_model([x_seg.cuda().float(), flow.cuda()])

            hd95  = 0
            count = 0
            assd  = 0
            fixed  = y_seg[0,0].detach().cpu().numpy()
            moving = x_seg[0,0].detach().cpu().numpy()
            moving_warped = def_out[0,0].detach().cpu().numpy()
            
            hd95  = hd95_cal(moving_warped, fixed, moving, labels=labels, vspace=np.ones(3)*2)#vspace depends on dataset
            assd  = assd_cal(moving_warped, fixed, moving, labels=labels, vspace=np.ones(3)*2)#
            
            
            tar = y.detach().cpu().numpy()[0, 0, :, :, :]
            jac_det = utils.jacobian_determinant_vxm(flow.detach().cpu().numpy()[0, :, :, :, :])
            eval_det.update(np.sum(jac_det <= 0) / np.prod(tar.shape), x.size(0))
            dsc_trans = utils.dice_abdo(def_out.long(), y_seg.long())
            dsc_raw   = utils.dice_abdo(x_seg.long(), y_seg.long())
            
            dis_f = flow.squeeze(0).movedim(0,-1)
            mask  = y_seg.squeeze(0).squeeze(0)

            non_diff_volume, non_diff_volume_percentage = 0,0
            # non_diff_volume, non_diff_volume_percentage = cal_ndv(dis_f, mask.detach().cpu().numpy(), remove_boundary=False)
            
            line = str(dsc_trans) + ',' + str(np.sum(jac_det <= 0) / np.prod(y_seg.shape))
            line = line  + ',' + str(hd95) + ',' + str(assd) + ',' + str(non_diff_volume)  + ',' + str(non_diff_volume_percentage)
                
            print('Trans dsc: {:.4f}, Raw dsc: {:.4f}'.format(dsc_trans.item(),dsc_raw.item()))
            eval_dsc_def.update(dsc_trans.item(), x.size(0))
            eval_dsc_raw.update(dsc_raw.item(), x.size(0))
            
            csv_writter(line, 'Quantitative_Results/' + model_folder[:-1])
            stdy_idx += 1
            
        print('Deformed DSC: {:.3f} +- {:.3f}, Affine DSC: {:.3f} +- {:.3f}'.format(eval_dsc_def.avg,
                                                                                    eval_dsc_def.std,
                                                                                    eval_dsc_raw.avg,
                                                                                    eval_dsc_raw.std))
        print('deformed det: {}, std: {}'.format(eval_det.avg, eval_det.std))

def csv_writter(line, name):
    with open(name+'.csv', 'a') as file:
        file.write(line)
        file.write('\n')
        
if __name__ == '__main__':
    '''
    GPU configuration
    '''
    GPU_iden = 0
    GPU_num = torch.cuda.device_count()
    print('Number of GPU: ' + str(GPU_num))
    for GPU_idx in range(GPU_num):
        GPU_name = torch.cuda.get_device_name(GPU_idx)
        print('     GPU #' + str(GPU_idx) + ': ' + GPU_name)
    torch.cuda.set_device(GPU_iden)
    GPU_avai = torch.cuda.is_available()
    print('Currently using: ' + torch.cuda.get_device_name(GPU_iden))
    print('If the GPU is available? ' + str(GPU_avai))
    main()

**Comment by x-xc** [comment]
Example test file.

**Comment by yingkewei** [comment]
非常感谢！
